tensorflow/python/keras/benchmarks/keras_examples_benchmarks/mnist_conv_custom_training_benchmark_test.py

```diff
@@ -0,0 +1,311 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Benchmarks using custom training loop on MNIST dataset."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import timeit
+import numpy as np
+
+import tensorflow as tf
+
+from tensorflow.python.keras.benchmarks import distribution_util
+
+class CustomMnistBenchmark(tf.test.Benchmark):
+  """Benchmarks for custom training loop using `tf.test.Benchmark`."""
+
+  def __init__(self):
+    super(CustomMnistBenchmark, self).__init__()
+    self.num_classes = 10
+    self.input_shape = (28, 28, 1)
+    self.epochs = 15
+    (x_train, y_train), _ = tf.keras.datasets.mnist.load_data()
+    x_train = x_train.astype('float32') / 255
+    x_train = np.expand_dims(x_train, -1)
+    y_train = tf.keras.utils.to_categorical(y_train, self.num_classes)
+    self.num_examples = x_train.shape[0]
+    #  Use `tf.data.Dataset` for custom training loop.
+    self.train_dataset = tf.data.Dataset.from_tensor_slices((x_train, y_train))
+
+  def _build_model(self):
+    """Model from https://keras.io/examples/vision/mnist_convnet/."""
+    model = tf.keras.Sequential([
+        tf.keras.Input(shape=self.input_shape),
+        tf.keras.layers.Conv2D(32, kernel_size=(3, 3), activation='relu'),
+        tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
+        tf.keras.layers.Conv2D(64, kernel_size=(3, 3), activation='relu'),
+        tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
+        tf.keras.layers.Flatten(),
+        tf.keras.layers.Dropout(0.5),
+        tf.keras.layers.Dense(self.num_classes, activation='softmax'),
+    ])
+
+    return model
+
+  def compute_loss(self,
+                   targets,
+                   predictions,
+                   loss_fn,
+                   batch_size):
+    """Compute average loss."""
+    per_example_loss = loss_fn(targets, predictions)
+    return tf.nn.compute_average_loss(
+        per_example_loss, global_batch_size=batch_size)
+
+  @tf.function
```

Comment:
You may enable `experimental_relax_shapes` for potential better performance: https://github.com/tensorflow/tensorflow/blob/387414fa3290211996230a9b2ed63cda22a14340/tensorflow/python/keras/engine/training.py#L789

Reply:
Yeah, I will keep it in mind.

Reply:
Can you add it in this PR? This is for consistency with built-in `fit`

Reply:
BTW, you mentioned there would be a fix of README in some PR? Should it be included in this PR?

Reply:
> BTW, you mentioned there would be a fix of README in some PR? Should it be included in this PR?

Yeah, I plan to update the README.md soon.

Reply:
> Can you add it in this PR? This is for consistency with built-in `fit`

Let me check the usage, should be update later.